src/sqlfluff/rules/L039.py

```diff
@@ -133,12 +141,86 @@ def _eval(self, context: RuleContext) -> Optional[List[LintResult]]:
                         violations.append(
                             LintResult(
                                 anchor=seg,
-                                fixes=[
-                                    LintFix.replace(
-                                        seg,
-                                        [WhitespaceSegment()],
-                                    )
-                                ],
+                                fixes=[LintFix.replace(seg, [WhitespaceSegment()])],
                             )
                         )
         return violations
+
+    def _skip_aliases(self, context: RuleContext, seg) -> bool:
+        """Checks whether segment formatting was handled by _align_aliases.
+
+        Args:
+            context: Context passed to _eval function.
+            seg: A whitespace segment.
+
+        Return:
+            bool: whether seg is part of a select clause element containing an alias.
+        """
+        segments = context.segment.segments
+        if context.segment.is_type("select_clause_element"):
+            segment_index = segments.index(seg)
+            # If seg is last segment, we know it cant be before an alias
+            if len(segments) > segment_index + 1:
+                prev_seg = segments[segment_index - 1]
+                next_seg = segments[segment_index + 1]
+                # Check to see if previous segment is column reference or expression
+                prev_is_col_expression = prev_seg.is_type(
+                    "expression"
+                ) or prev_seg.is_type("column_reference")
+                # Check to see if next segment is alias expression
+                next_is_alias = next_seg.is_type("alias_expression")
+                if prev_is_col_expression and next_is_alias:
+                    return True
+        return False
+
+    def _pad_unaligned_aliases(self, elements, max_len) -> List[LintFix]:
+        """Finds expressions before aliases, and ensures they are padded to line up."""
+        fixes = []
+        # Loop over `select_clause_element`s again to pad each expression/apply fixes
+        for element in elements:
+            if element.is_type("select_clause_element"):
+                for expression_segment in element.segments:
+                    is_expression = expression_segment.is_type("expression")
+                    is_column = expression_segment.is_type("column_reference")
+                    if is_expression or is_column:
+                        # Determine how much padding is needed for expression
+                        padding = max_len - expression_segment.matched_length + 1
+                        # Fetch existing WhiteSpace element following this expression
+                        old_white_space = element.segments[
```

Comment:
I ran the rule on some SQL files at work, and I'm seeing a number of occurrences of this exception:
```
[L039] Applying rule L039 threw an Exception: tuple index out of range
Traceback (most recent call last):
  File "/Users/barry.hart/dev/sqlfluff/src/sqlfluff/core/rules/base.py", line 497, in crawl
    res = self._eval(context=context)
  File "/Users/barry.hart/dev/sqlfluff/src/sqlfluff/rules/L039.py", line 89, in _eval
    align_violation = self._align_aliases(context)
  File "/Users/barry.hart/dev/sqlfluff/src/sqlfluff/rules/L039.py", line 216, in _align_aliases
    fixes = self._pad_unaligned_aliases(
  File "/Users/barry.hart/dev/sqlfluff/src/sqlfluff/rules/L039.py", line 189, in _pad_unaligned_aliases
    old_white_space = element.segments[
IndexError: tuple index out of range
```

I'll try and get a test case.

Reply:
Here's one test case:
```
WITH categories AS (
    SELECT
    tbl1.code category_code
    from tbl1)
SELECT category_code
FROM categories
WHERE category_code is not null
```

Reply:
Another one:
```
SELECT
    a,
    b
FROM tbl
```